### PR TITLE
[Rust] Read Partially from a plain encoded batch

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/encodings/plain.rs [hunk 1]
```diff
@@ -120,24 +120,32 @@ impl<'a> PlainDecoder<'a> {
...
+    /// Decode primitive values, from "offset" to "offset + length".
+    ///
+    async fn decode_primitive(&self, start: usize, end: usize) -> Result<ArrayRef> {
+        if end > self.length {
+            return Err(Error::IO(format!(
+                "PlainDecoder: request([{}..{}]) out of range: [0..{}]",
+                start, end, self.length
+            )));
+        }
+        let start_offset = match self.data_type {
```
Comment (changhiskhan) [inline rust/src/encodings/plain.rs:132]: Let’s make a “index_to_offset” function?

File: rust/src/encodings/plain.rs [hunk 2]
```diff
@@ -120,24 +120,32 @@ impl<'a> PlainDecoder<'a> {
...
+            _ => get_primitive_byte_width(self.data_type)? * start,
+        };
+        let end_offset = match self.data_type {
+            DataType::Boolean => bit_util::ceil(end, 8),
+            _ => get_primitive_byte_width(self.data_type)? * end,
         };
         let range = Range {
-            start: self.position,
-            end: self.position + array_bytes,
+            start: self.position + start_offset,
```
Comment (changhiskhan) [inline rust/src/encodings/plain.rs:141]: I vaguely remember that the C++ implementation was more complicated than this. Any edge cases that’s not covered here?

File: rust/src/encodings/plain.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/encodings/plain.rs]: done

File: rust/src/encodings/plain.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/encodings/plain.rs:144]: This does not cover the logic in C++ `PlainEncoder::Take(indices)` yet. 
For C++, `PlainEncoder::ToArray(...)` this PR reaches the parity.
https://github.com/eto-ai/lance/blob/main/cpp/src/lance/encodings/plain.cc#L127-L149

Added 2 more tests to cover edge cases nonetheless.

**Review by changhiskhan — APPROVED** [review]
